zpages/tracez_test.go

```diff
@@ -0,0 +1,218 @@
+			wantContains: []string{"non-existing"},
+		},
+		{
+			name:         "latency spans (type=1)",
+			url:          "/tracez?zspanname=completed-span&ztype=1",
+			wantStatus:   http.StatusOK,
+			wantContains: []string{},
+		},
+		{
+			name:         "error spans (type=2)",
+			url:          "/tracez?zspanname=error-span&ztype=2",
+			wantStatus:   http.StatusOK,
+			wantContains: []string{},
+		},
+		{
+			name:       "with latency bucket",
+			url:        "/tracez?zspanname=completed-span&ztype=1&zlatencybucket=0",
+			wantStatus: http.StatusOK,
+		},
+		{
+			name:         "all parameters",
+			url:          "/tracez?zspanname=completed-span&ztype=1&zlatencybucket=2",
+			wantStatus:   http.StatusOK,
+			wantContains: []string{},
+		},
+		{
+			name:       "type only",
+			url:        "/tracez?ztype=1",
+			wantStatus: http.StatusOK,
+		},
+		{
+			name:       "latency bucket only",
+			url:        "/tracez?zlatencybucket=3",
+			wantStatus: http.StatusOK,
+		},
+		{
+			name:       "invalid type value",
+			url:        "/tracez?ztype=invalid",
+			wantStatus: http.StatusOK,
+		},
+		{
+			name:       "negative latency bucket",
+			url:        "/tracez?zlatencybucket=-1",
+			wantStatus: http.StatusOK,
+		},
+		{
+			name:         "multiple spans with same name",
+			url:          "/tracez?zspanname=multi-span&ztype=1",
+			wantStatus:   http.StatusOK,
+			wantContains: []string{"multi-span"},
+		},
+		{
+			name:       "invalid form encoding",
+			url:        "/tracez?%zzz",
+			wantStatus: http.StatusBadRequest,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			req := httptest.NewRequest(http.MethodGet, tt.url, http.NoBody)
+			w := httptest.NewRecorder()
+
+			handler.ServeHTTP(w, req)
+
+			resp := w.Result()
+			defer resp.Body.Close()
+
+			assert.Equal(t, tt.wantStatus, resp.StatusCode)
+
+			// Verify content type for successful responses
+			if resp.StatusCode == http.StatusOK && tt.url == "/tracez" {
+				assert.Equal(t, "text/html; charset=utf-8", resp.Header.Get("Content-Type"))
+			}
+
+			body := w.Body.String()
+
+			// Only check body content for successful responses
+			if resp.StatusCode == http.StatusOK {
+				assert.NotEmpty(t, body, "expected non-empty response body")
+
+				for _, want := range tt.wantContains {
+					assert.Contains(t, body, want)
+				}
+				for _, notWant := range tt.wantNotContain {
+					assert.NotContains(t, body, notWant)
+				}
+			}
+		})
+	}
+}
+
+func TestTracezHandler_ConcurrentRequests(t *testing.T) {
+	sp := NewSpanProcessor()
+	defer func() {
+		require.NoError(t, sp.Shutdown(t.Context()))
+	}()
+
+	tp := sdktrace.NewTracerProvider(
+		sdktrace.WithSpanProcessor(sp),
+	)
+	defer func() {
+		require.NoError(t, tp.Shutdown(t.Context()))
+	}()
+
+	tracer := tp.Tracer("test-tracer")
+	ctx := t.Context()
+
+	for range 10 {
+		_, span := tracer.Start(ctx, "concurrent-span")
+		span.End()
+	}
+
+	handler := NewTracezHandler(sp)
+
+	done := make(chan bool, 5)
```

Comment:
Rather than using a channel, it seems using a mutex would be clearer.
See https://github.com/open-telemetry/opentelemetry-go-contrib/blob/6342fe6da91e0e5d126ac69774dba3ccc3cf559a/propagators/autoprop/registry_test.go#L42 for an example.